python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java

```diff
@@ -282,6 +306,62 @@ void test_optional_type_annotations() {
       .containsExactlyInAnyOrder("typing.Optional");
   }
 
+  @Test
+  void test_union_type_annotation_with_builtins() {
+    DeclaredType union = new DeclaredType(new SymbolImpl("Union", "typing.Union"), List.of(new DeclaredType(BuiltinTypes.INT), new DeclaredType(BuiltinTypes.STR)));
+    assertThat(lastExpression(
+      "def f(x: int | str):",
+      "  x").type()).isEqualTo(union);
+  }
+
+  @Test
+  void test_union_type_annotation_inside_generics() {
+    DeclaredType union = new DeclaredType(new SymbolImpl("Union", "typing.Union"), List.of(new DeclaredType(BuiltinTypes.INT), new DeclaredType(BuiltinTypes.STR)));
+    DeclaredType listOfUnion = new DeclaredType(new SymbolImpl(BuiltinTypes.LIST, BuiltinTypes.LIST), List.of(union));
+    assertThat(lastExpression(
+      "def f(x: list[int | str]):",
+      "  x").type()).isEqualTo(listOfUnion);
+  }
+
+  @Test
+  void test_union_type_annotation_with_unknown_types() {
+    assertThat(lastExpression(
+      "def f(x: int | A):",
+      "  x").type()).isEqualTo(anyType());
+
+    assertThat(lastExpression(
+      "def f(x: B | int):",
+      "  x").type()).isEqualTo(anyType());
+  }
+
+  @Test
+  void test_union_type_annotation_with_any_type() {
+    assertThat(lastExpression(
+      "from typing import Any",
+      "def f(x: Any):",
+      "  x").type()).isEqualTo(anyType());
+  }
+
```

Comment:
The following case seems to fail:
```java
  @Test
  void test_declared_type_outside_of_method() {
    DeclaredType union = new DeclaredType(new SymbolImpl("Union", "typing.Union"), List.of(new DeclaredType(BuiltinTypes.INT),
      new DeclaredType(BuiltinTypes.STR)));
    assertThat(lastExpression(
      "class Foo():",
      "  x: int | str",
      "  x").type()).isEqualTo(union);
  }
```
I am not entirely sure that it is within the scope of this ticket, but maybe it would be worth investigating?

Reply:
Ok so I have taken a look how I could test this. It is a bit complicated as we cannot use `lastExpression`. Also as discussed this yet out of scope so for now we can leave it as is and come back to it later on if needed.

Reply:
Probably the class attributes ticket would also be a blocker for this (although I'm not sure it would work in the current state of things once both tickets are merged).

A way to rewrite the test could make use of a separate function that instantiates the class and then accesses the attribute (similar to the tests in this [PR](https://github.com/SonarSource/sonar-python/pull/1653)) because currently the fact there is no function fails the test before it can actually test anything.

I think it makes sense to have such a test (either in this or in the other PR) to ensure we eventually cover this case though.